main_test.go

```diff
@@ -0,0 +1,282 @@
+		t.Fatal(handleErr(err))
+	}
+}
+
+func setupClusterObject(clusterName, namespace string) error {
+	out, err := capdctl("cluster", "--cluster-name", clusterName, "--namespace", namespace)
+	if err != nil {
+		return err
+	}
+	return pipeToKubectlApply(bytes.NewReader(out))
+}
+func setupControlPlaneObject(clusterName, namespace, version string) error {
+	out, err := capdctl(
+		"control-plane",
+		"--name", controlPlaneName(clusterName),
+		"--namespace", namespace,
+		"--cluster-name", clusterName,
+		"--version", version)
+	if err != nil {
+		return err
+	}
+	return pipeToKubectlApply(bytes.NewReader(out))
+}
+func setupWorkerObject(clusterName, namespace, version string) error {
+	out, err := capdctl("worker",
+		"--name", workerName(clusterName),
+		"--namespace", namespace,
+		"--cluster-name", clusterName,
+		"--version", version)
+	if err != nil {
+		return err
+	}
+	return pipeToKubectlApply(bytes.NewReader(out))
+}
+
+// controlPlaneName generates the base control-plane node name for a given cluster
+func controlPlaneName(clusterName string) string {
+	return fmt.Sprintf("%s-control-plane", clusterName)
+}
+
+// workerName generates the base worker node name for a given cluster
+func workerName(clusterName string) string {
+	return fmt.Sprintf("%s-worker", clusterName)
+}
+
+// secretName generates the name of the kubeconfig secret stored on the management cluster
+func secretName(clusterName string) string {
+	return fmt.Sprintf("kubeconfig-%s", clusterName)
+}
+
+func handleErr(err error) string {
+	if e, ok := err.(*exec.ExitError); ok {
+		return string(e.Stderr)
+	}
+	return err.Error()
+}
+
+func setupManagementCluster(t *testing.T) {
+	t.Helper()
+	// if a cluster named kind already exists then we assume we're good to go:
+	// TODO: this isn't very good because we have to clean up the management cluster or use unique machines etc.
+	clusters, err := kind("get", "clusters")
+	if err != nil {
+		panic(err)
+	}
+	for _, cluster := range bytes.Split(clusters, []byte("\n")) {
+		if string(bytes.TrimSpace(cluster)) == "kind" {
+			fmt.Println("Detected a management cluster.")
+			return
+		}
+	}
+
+	if _, err := capdctl("setup"); err != nil {
+		t.Fatal(handleErr(err))
+	}
+	if err := installCRDs(); err != nil {
+		t.Fatal(handleErr(err))
+	}
+	if err := installCAPD(); err != nil {
+		t.Fatal(handleErr(err))
+	}
+}
+
+func pipeToKubectlApply(reader io.Reader) error {
+	_, err := kubectl(reader, "kind", "apply", "-f", "-")
+	if err != nil {
+		return err
+	}
+	return nil
+}
+
+func installCAPD() error {
+	out, err := capdctl("capd")
+	if err != nil {
+		return err
+	}
+	return pipeToKubectlApply(bytes.NewReader(out))
+}
+
+func installCRDs() error {
+	out, err := capdctl("crds")
+	if err != nil {
+		return err
+	}
+	return pipeToKubectlApply(bytes.NewReader(out))
+}
+
+func capdctl(args ...string) ([]byte, error) {
+	cmd := exec.Command("capdctl", args...)
+	return cmd.Output()
+}
+
+// TODO this input is not right...
```

Comment:
Yeah, something like kubectl().ForCluster(c).WithReader(r).Run("apply", "-f", "-") might be nice?